### PR TITLE
When a user add a new announcement the company and the city must be connected

### PR DESCRIPTION
**Pull request by mmarchois**
PR 15

See https://github.com/poledev/mathieu/issues/7

### CODE REVIEW

File: src/SensioLabs/JobBoardBundle/Controller/BackendController.php [hunk 1]
```diff
@@ -61,6 +61,7 @@ public function editAction(Request $request, Job $job)
             'attr' => ['novalidate' => 'novalidate'],
             'is_admin' => true,
             'validation_groups' => ['Default', 'Backend'],
+            'em' => $this->getDoctrine()->getManager(),
```
Comment (ogizanagi) [inline src/SensioLabs/JobBoardBundle/Controller/BackendController.php]: In order to inject dependencies to your FormType, you must declare it as a service and inject it through the constructor.
Dependencies should not be options.

File: src/SensioLabs/JobBoardBundle/Controller/BaseController.php [hunk 2]
```diff
@@ -78,6 +78,7 @@ public function postAction(Request $request)
         $form = $this->createForm(new JobType(), $job, [
             'action' => $this->generateUrl('job_post'),
             'attr'   => ['novalidate' => 'novalidate'],
+            'em' => $this->getDoctrine()->getManager(),
```
Comment (ogizanagi) [inline src/SensioLabs/JobBoardBundle/Controller/BaseController.php]: Same here

File: src/SensioLabs/JobBoardBundle/Controller/JobController.php [hunk 3]
```diff
@@ -68,11 +68,13 @@ public function updateAction(Request $request, Job $job)
...
 
         $form = $this->createForm(new JobType(), $job, [
-            'attr' => ['novalidate' => 'novalidate']
+            'attr' => ['novalidate' => 'novalidate'],
+            'em' => $this->getDoctrine()->getManager(),
         ]);
 
         if ($form->handleRequest($request)->isValid()) {
-            $this->getDoctrine()->getManager()->flush($job);
+            $this->getDoctrine()->getManager()->flush([$job, $job->getCompany()]);
```
Comment (ogizanagi) [inline src/SensioLabs/JobBoardBundle/Controller/JobController.php]: Why not:
```diff
- $this->getDoctrine()->getManager()->flush([$job, $job->getCompany()]);
+ $this->getDoctrine()->getManager()->flush($job);
```

?

File: src/SensioLabs/JobBoardBundle/DataFixtures/ORM/LoadJobsData.php [hunk 4]
```diff
@@ -23,9 +23,7 @@ public function load(ObjectManager $manager)
         foreach ($yml as $value) {
             $job = new Job();
             $job->setContract($value['contract']);
-            $job->setCompany($value['company']);
-            $job->setCity($value['city']);
-            $job->setCountry($value['country']);
+            $job->setCompany($this->getReference('SensioLabs_FR'));
```
Comment (ogizanagi) [inline src/SensioLabs/JobBoardBundle/DataFixtures/ORM/LoadJobsData.php]: I think it would be best if your yml files entirely reflect your data. IE: the company field must be present in your yaml file.

File: src/SensioLabs/JobBoardBundle/DataFixtures/YML/jobs.yml [hunk 5]
```diff
@@ -1,81 +1,54 @@
 -
     title: Developer symfony - 1
     contract: cdd
-    company: SensioLabs
```
Comment (ogizanagi) [inline src/SensioLabs/JobBoardBundle/DataFixtures/YML/jobs.yml]: Here. Define the company reference name.

File: src/SensioLabs/JobBoardBundle/Entity/Company.php [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+/**
+ * @ORM\Entity()
+ * @ORM\Table(
+ *      name = "company",
+ *      uniqueConstraints = {
+ *          @ORM\UniqueConstraint(name="company_idx", columns={"city", "country", "name"})
+ *      }
+ * )
+ */
+class Company
```
Comment (ogizanagi) [inline src/SensioLabs/JobBoardBundle/Entity/Company.php]: `UniqueEntityConstraint` ?

File: src/SensioLabs/JobBoardBundle/Entity/Company.php [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+{
+    /**
+     * @ORM\Id
+     * @ORM\Column(type = "integer")
+     * @ORM\GeneratedValue(strategy = "AUTO")
+     *
+     * @var integer
+     */
+    protected $id;
```
Comment (ogizanagi) [inline src/SensioLabs/JobBoardBundle/Entity/Company.php]: private

File: src/SensioLabs/JobBoardBundle/Entity/Company.php [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+     * @var string
+     */
+    protected $country;
+
+    /**
+     * @param string $name
+     * @param string $city
+     * @param string $country
+     */
+    public function __construct($name = null, $city= null, $country = null)
```
Comment (ogizanagi) [inline src/SensioLabs/JobBoardBundle/Entity/Company.php]: Why do you authorize `null` values ?

File: src/SensioLabs/JobBoardBundle/Controller/BackendController.php (hunk 1 above)
Comment (mmarchois) [inline src/SensioLabs/JobBoardBundle/Controller/BackendController.php]: In the symfony documentation, the entity manager is inject by form options http://symfony.com/doc/current/cookbook/form/data_transformers.html#using-the-transformer

File: src/SensioLabs/JobBoardBundle/Entity/Job.php [hunk 9]
```diff
@@ -114,7 +89,7 @@ class Job
     /**
      * @var User
      *
-     * @ORM\ManyToOne(targetEntity = "SensioLabs\JobBoardBundle\Entity\User")
+     * @ORM\ManyToOne(targetEntity = "SensioLabs\JobBoardBundle\Entity\User", cascade={"persist"})
```
Comment (ogizanagi) [inline src/SensioLabs/JobBoardBundle/Entity/Job.php]: Is it related to this PR ?
Why do you need it ?

File: src/SensioLabs/JobBoardBundle/Entity/Job.php [hunk 10]
```diff
@@ -221,9 +203,9 @@ public function getCompany()
     }
 
     /**
-     * @param string $company The blog company
+     * @param Company $company aThe job company
```
Comment (ogizanagi) [inline src/SensioLabs/JobBoardBundle/Entity/Job.php]: typo (`aThe`)

File: src/SensioLabs/JobBoardBundle/Entity/Job.php (hunk 9 above)
Comment (mmarchois) [inline src/SensioLabs/JobBoardBundle/Entity/Job.php]: Not related to this PR, i forgot to remove it

File: src/SensioLabs/JobBoardBundle/Controller/BaseController.php [hunk 11]
```diff
@@ -75,7 +74,7 @@ public function postAction(Request $request)
         $job = new Job();
         $job->setUser($this->getUser());
 
-        $form = $this->createForm(new JobType(), $job, [
+        $form = $this->createForm($this->get('job_board.form.job'), $job, [
```
Comment (ogizanagi) [inline src/SensioLabs/JobBoardBundle/Controller/BaseController.php]: That's not how to use a form type :stuck_out_tongue_winking_eye: 
Use the alias you define in the form.type tag directly. No need to search for it in the container.

File: src/SensioLabs/JobBoardBundle/Resources/config/services.xml [hunk 12]
```diff
@@ -32,5 +32,8 @@
         <service id="job_board.voter.job" class="SensioLabs\JobBoardBundle\Security\Authorization\Voter\JobVoter">
             <tag name="security.voter" />
         </service>
+        <service id="job_board.form.job" class="SensioLabs\JobBoardBundle\Form\Type\JobType">
+            <argument type="service" id="doctrine.orm.entity_manager" />
```
Comment (ogizanagi) [inline src/SensioLabs/JobBoardBundle/Resources/config/services.xml]: Well, no I understand. You miss the `form.type` tag => http://symfony.com/doc/current/cookbook/form/create_custom_field_type.html#creating-your-field-type-as-a-service

File: src/SensioLabs/JobBoardBundle/Form/Type/JobType.php [hunk 13]
```diff
@@ -23,28 +39,12 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                     'placeholder' => 'job.placeholders.title'
                 ],
             ])
-            ->add('company', 'text', [
-                'label' => false,
-                'attr' => [
-                    'placeholder' => 'job.placeholders.company'
-                ],
-            ])
+            ->add('company', new CompanyType())->addModelTransformer($transformer)
```
Comment (ogizanagi) [inline src/SensioLabs/JobBoardBundle/Form/Type/JobType.php]: So, this ModelTransformer is applied on the `CompanyType` right ?
What will be the data the transformer will receive ?

File: src/SensioLabs/JobBoardBundle/Form/Transformer/JobCompanyTransformer.php [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+        $company = $this->om->getRepository('SensioLabsJobBoardBundle:Company')->findOneBy([
+            'name' => $jobCompany->getName(),
+            'city' => $jobCompany->getCity(),
+            'country' => $jobCompany->getCountry(),
+        ]);
+
+        // In case of insert
+        if (null === $jobCompany->getId()) {
+            if (null !== $company) {
+                $job->setCompany($company);
```
Comment (ogizanagi) [inline src/SensioLabs/JobBoardBundle/Form/Transformer/JobCompanyTransformer.php]: I just realized this is wrong. @salahm was totally right: this is not the responsibility of the transformer. This is particularly visible as you use mutators to **replace** underlying model data instead of transforming the value the transformer received into a model representation.
You should do that using FormEvents IMO.

File: src/SensioLabs/JobBoardBundle/Controller/BaseController.php (hunk 11 above)
Comment (florianbelhomme) [inline src/SensioLabs/JobBoardBundle/Controller/BaseController.php]: +1

File: src/SensioLabs/JobBoardBundle/DataFixtures/ORM/LoadCompaniesData.php [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+use Symfony\Component\Yaml\Yaml;
+
+class LoadCompaniesData extends AbstractFixture implements OrderedFixtureInterface
+{
+    /**
+     * {@inheritDoc}
+     */
+    public function load(ObjectManager $manager)
+    {
+        $yml = Yaml::parse(__DIR__ . '/../YML/companies.yml');
```
Comment (florianbelhomme) [inline src/SensioLabs/JobBoardBundle/DataFixtures/ORM/LoadCompaniesData.php]: __DIR__ should be banned from your devs, you have remove it already and Insight will cry

File: src/SensioLabs/JobBoardBundle/Form/Type/JobType.php [hunk 16]
```diff
@@ -9,11 +11,26 @@
 class JobType extends AbstractType
 {
     /**
+     * @var EntityManagerInterface
+     */
+    protected $entityManager;
```
Comment (florianbelhomme) [inline src/SensioLabs/JobBoardBundle/Form/Type/JobType.php]: OM / EM : tu est inconsistent dans ton nommage/utilisation

File: src/SensioLabs/JobBoardBundle/Form/Type/JobType.php [hunk 17]
```diff
@@ -9,11 +11,26 @@
...
+        $this->entityManager = $entityManager;
+    }
+
+    /**
      * @param FormBuilderInterface $builder The builder
      * @param array                $options The options
      */
     public function buildForm(FormBuilderInterface $builder, array $options)
     {
+        $transformer = new JobCompanyTransformer($this->entityManager);
```
Comment (florianbelhomme) [inline src/SensioLabs/JobBoardBundle/Form/Type/JobType.php]: dans ton transformer tu l'appelle OM

Comment (mmarchois) [inline src/SensioLabs/JobBoardBundle/Form/Type/JobType.php]: Je change ca

File: src/SensioLabs/JobBoardBundle/Form/Transformer/JobCompanyTransformer.php (hunk 14 above)
Comment (mmarchois) [inline src/SensioLabs/JobBoardBundle/Form/Transformer/JobCompanyTransformer.php]: I use form event PRE_SUBMIT instead

File: src/SensioLabs/JobBoardBundle/Resources/config/services.xml (hunk 12 above)
Comment (mmarchois) [inline src/SensioLabs/JobBoardBundle/Resources/config/services.xml]: I added the form.type alias

File: src/SensioLabs/JobBoardBundle/Form/Type/JobType.php [hunk 18]
```diff
@@ -83,6 +84,33 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
+                'name' => $jobCompany['name'],
+                'city' => $jobCompany['city'],
+                'country' => $jobCompany['country'],
+            ]);
+
+            if (null === $company) {
+                $company = new Company();
+                $company->setCity($jobCompany['city']);
+                $company->setName($jobCompany['name']);
+                $company->setCountry($jobCompany['country']);
```
Comment (ogizanagi) [inline src/SensioLabs/JobBoardBundle/Form/Type/JobType.php]: I have doubts: Are those three lines above (calling mutators) needed ? Won't it be replaced anyway during submit by the data in the request ? (What you have in `$event->getData()['company']`)
https://github.com/symfony/symfony/blob/2aeaa22e3171b5e1f10f2cc2a7ff7f6608b65f9d/src/Symfony/Component/Form/Form.php#L552-L597

File: src/SensioLabs/JobBoardBundle/Resources/config/services.xml [hunk 19]
```diff
@@ -32,5 +32,9 @@
...
+            <tag name="form.type" alias="sensiolabs_jobboardbundle_job" />
```
Comment (ogizanagi) [inline src/SensioLabs/JobBoardBundle/Resources/config/services.xml]: IMO, you don't need such a complex name for your own form types, unless there is an actual conflict issue.
Even third party bundles keep their form types aliases very simple and prefix them with vendor name only when necessary.

File: src/SensioLabs/JobBoardBundle/Controller/BaseController.php [hunk 20]
```diff
@@ -75,7 +74,7 @@ public function postAction(Request $request)
         $job = new Job();
         $job->setUser($this->getUser());
 
-        $form = $this->createForm(new JobType(), $job, [
+        $form = $this->createForm('sensiolabs_jobboardbundle_job', $job, [
             'action' => $this->generateUrl('job_post'),
```
Comment (ogizanagi) [inline src/SensioLabs/JobBoardBundle/Controller/BaseController.php]: Why did you need to set the `action` explicitly here ?

Comment (mmarchois) [inline src/SensioLabs/JobBoardBundle/Controller/BaseController.php]: I got used to always specify it. I can remove it

File: src/SensioLabs/JobBoardBundle/Form/Type/JobType.php (hunk 18 above)
Comment (mmarchois) [inline src/SensioLabs/JobBoardBundle/Form/Type/JobType.php]: The $event->getData()['company'] method return an array, i can use it directly

File: src/SensioLabs/JobBoardBundle/Resources/config/services.xml (hunk 19 above)
Comment (mmarchois) [inline src/SensioLabs/JobBoardBundle/Resources/config/services.xml]: I changed it

File: src/SensioLabs/JobBoardBundle/Form/Type/JobType.php (hunk 18 above)
Comment (ogizanagi) [inline src/SensioLabs/JobBoardBundle/Form/Type/JobType.php]: That's not what I meant. You're already using it. What I meant is that you do not need to set the `Company` instance properties from what you have in the request, as it is already handled during submit.
But I might be wrong, and forgot something.
However, try to comment l.105 to l.107.

Comment (mmarchois) [inline src/SensioLabs/JobBoardBundle/Form/Type/JobType.php]: I'm going to try this, i'll tell you

Comment (mmarchois) [inline src/SensioLabs/JobBoardBundle/Form/Type/JobType.php]: It work's without setter
